keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md

```diff
@@ -0,0 +1,1392 @@
+    *   Per Type: Migrate all validation rules for a single API. e.g. all validation rules of `ReplicationControllerSpec`.` `Df
+    *   Per Group (not recommended): Migrate the entire API group/version.
+*   Label tasks appropriately.
+2. Community-Driven API Migration:
+*   Community:
+    *   Analyze existing handwritten validation.
+    *   Add appropriate IDL tags to API type definitions.
+    *   Run `validation-gen` to generate validation code.
+    *   Update unit tests to use the generated validation and ensure coverage.
+    *   Update E2E tests to verify behavior with declarative validation.
+    *   Submit pull requests (PRs) with the changes.
+*   Core Team:
+    *   Provide technical guidance and support to community contributors.
+    *   Review PRs.
+    *   Monitor the tracking issue and adjust the plan as needed.
+3. Using Schemas for Validation (Joint Effort):
+*   Core Team:
+    *   Enable validation through generated schemas for migrated resources (controlled by DeclarativeValidation feature gate).
+    *   Implement logic to populate default values from schemas.
+*   Community:
+    *   Run E2E tests with declarative validation enabled.
+
+
+#### 	Phase3: Finalization and GA (Core Team and community)
+
+
+
+1. After DeclarativeValidation reaches GA
+2. Deprecation of Legacy Validation is Announced
+3. ...Deprecation wait period passes (period adhering to community policy)…
+4. Legacy validation code that is being validated declaratively can safely be deleted
+
+### Tagging and Validating Against Internal vs Versioned Types
+
+
+In Kubernetes there are internal schema representations and versioned schema representations for k8s types. The IDL tags can be added to either set of types (and plumbed to validate against that type).  Below are the options and associated tradeoffs related to `validation-gen` using internal, versioned, or relevant combinations of the two:
+
+1. **Use versioned-type validation (Recommended)**.  Tags would be added on every version’s definition of a given type or field.
+    *   Pro: Validation is obvious for a given version by looking at types.go.
+    *   Con: Tags across versions need to stay in sync.
+        *   Rebuttal: This can be easily solved with tests.
+    *   Pro: Field-path differences between API versions are handled.
+    *   Con: Requires we deal with the performance implications up front.
+        *   Don’t want to convert internal &lt;-> versioned over and over.
+2. **Use internal-type validation**.  Tags would be added on the “unversioned” definition of a given type or field.
+    *   Pro: No immediate performance implications.
+    *   Con: Need to also add ~equivalent tags to versioned types, in order to generate openapi. Tags across versions need to stay in sync (manually or with automation).
+        *   Harder to automate cross-checking.
+    *   Con: Need to handle field-path differences between versions (e.g. with post processing).
```

Comment:
Con: closely-related tags like +default are already defined on versioned types, which means we need to "thread together" fields across internal and versioned, even across renames or refactoring.

Reply:
Added the suggested `Con`